dotcom-rendering/src/client/userFeatures/user-features-lib.ts

```diff
@@ -0,0 +1,179 @@
+import { getCookie, setCookie } from '@guardian/libs';
+
+// from cookies
+
+const timeInDaysFromNow = (daysFromNow: number): string => {
+	const tmpDate = new Date();
+	tmpDate.setDate(tmpDate.getDate() + daysFromNow);
+	return tmpDate.getTime().toString();
+};
+
+const cookieIsExpiredOrMissing = (cookieName: string): boolean => {
+	const expiryDateFromCookie = getCookie({ name: cookieName });
+	if (!expiryDateFromCookie) return true;
+	const expiryTime = parseInt(expiryDateFromCookie, 10);
+	const timeNow = new Date().getTime();
+	return timeNow >= expiryTime;
+};
+
+// from manage-ad-free-cookies
+
+const AD_FREE_USER_COOKIE = 'GU_AF1';
+
+const getAdFreeCookie = (): string | null =>
+	getCookie({ name: AD_FREE_USER_COOKIE });
+
+const adFreeDataIsPresent = (): boolean => {
+	const cookieVal = getAdFreeCookie();
+	if (!cookieVal) return false;
+	return !Number.isNaN(parseInt(cookieVal, 10));
+};
+
+/*
+ * Set the ad free cookie
+ *
+ * @param daysToLive - number of days the cookie should be valid
+ */
+const setAdFreeCookie = (daysToLive = 1): void => {
+	const expires = new Date();
+	expires.setMonth(expires.getMonth() + 6);
+	setCookie({
+		name: AD_FREE_USER_COOKIE,
+		value: expires.getTime().toString(),
+		daysToLive,
+	});
+};
+
+// from lib/utils/fetch-json
+
+/**
+ * Check that path is a path-absolute-URL string as described in https://url.spec.whatwg.org/#path-absolute-url-string
+ * A path-absolute-URL string is U+002F (/) followed by a path-relative-URL string, for instance `/plop` or `/plop/plop`
+ */
+function isPathAbsoluteURL(path: string): boolean {
+	return !RegExp('^(https?:)?//').exec(path);
+}
+
+const fetchJson = async (
+	resource: string,
+	init: RequestInit = {},
+): Promise<unknown> => {
+	if (typeof resource !== 'string') {
+		throw new Error('First argument should be of type `string`');
+	}
+
+	let path = resource;
+	if (isPathAbsoluteURL(path)) {
+		path = window.guardian.config.page.ajaxUrl + resource;
+		init.mode = 'cors';
+	}
+
+	const resp = await fetch(path, init);
+	if (resp.ok) {
+		switch (resp.status) {
+			case 204:
+				return {};
+			default:
+				try {
+					return resp.json();
+				} catch (ex) {
+					throw new Error(
+						`Fetch error while requesting ${path}: Invalid JSON response`,
+					);
+				}
+		}
+	}
+	throw new Error(`Fetch error while requesting ${path}: ${resp.statusText}`);
+};
+
+// from types/dates
+
+const dates = {
+	1: '01',
+	2: '02',
+	3: '03',
+	4: '04',
+	5: '05',
+	6: '06',
+	7: '07',
+	8: '08',
+	9: '09',
+	10: '10',
+	11: '11',
+	12: '12',
+	13: '13',
+	14: '14',
+	15: '15',
+	16: '16',
+	17: '17',
+	18: '18',
+	19: '19',
+	20: '20',
+	21: '21',
+	22: '22',
+	23: '23',
+	24: '24',
+	25: '25',
+	26: '26',
+	27: '27',
+	28: '28',
+	29: '29',
+	30: '30',
+	31: '31',
+} as const;
```

Comment:
Is this necessary to ship to users?

Reply:
A lot of these changes are being directly ported from the commercial bundle (in this case from our [`dates` module](https://github.com/guardian/commercial/blob/main/src/types/dates.ts)). While there may be other ways of achieving this, it would be good to get the code extraction out of the way first, so supporter revenue teams are no longer dependent on the commercial bundle.

Happy to chat offline about alternative solutions to this module 🙂

Reply:
A 20% increasing in the size of our `index.js` from 38kB to 45kB is non-trivial, so it would be nice to understand what follow-up mitigation tasks you have in mind. I appreciate that this may be code that is already shipped to clients, but by adding its code directly to this repository, it becomes a direct concern of @guardian/dotcom-platform.

Reply:
This may be a necessity [based on the priority of this script](https://github.com/guardian/dotcom-rendering/pull/9345/files#r1374540309). We need to be aware that this will increase the download and parsing time of the initial chunk, but it may be the most performant thing to do.

We’ve discussed this offline and we can accept this not being a blocker if there is some remediation plan in place, e.g. ensuring that this does not have a negative impact on web vitals.